WalletWasabi/Blockchain/Analysis/BlockchainAnalyzer.cs

```diff
@@ -3,60 +3,188 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Text;
+using WalletWasabi.Blockchain.Keys;
 using WalletWasabi.Blockchain.Transactions;
 
 namespace WalletWasabi.Blockchain.Analysis
 {
 	public class BlockchainAnalyzer
 	{
-		public BlockchainAnalyzer(int privacyLevelThreshold, Money dustThreshold)
+		public BlockchainAnalyzer(int privacyLevelThreshold)
 		{
 			PrivacyLevelThreshold = privacyLevelThreshold;
-			DustThreshold = dustThreshold;
 		}
 
 		public int PrivacyLevelThreshold { get; }
-		public Money DustThreshold { get; }
 
 		/// <summary>
 		/// Sets clusters and anonymity sets for related HD public keys.
 		/// </summary>
 		public void Analyze(SmartTransaction tx)
 		{
-			// Dust attack could ruin the analysis, so it's better not to incorporate this.
-			// For example if someone would dust a mixed coin of ours, its anon score would end up being 1,
-			// even though the duster does not know who it sent the money.
-			// FTR at the time of writing this comment the transaction processor doesn't even let dust to come this far,
-			// but it's better to be prepared for future changes if it's not too computationally expensive.
-			foreach (var newCoin in tx.WalletOutputs.Where(x => x.Amount > DustThreshold))
+			var inputCount = tx.Transaction.Inputs.Count;
+			var outputCount = tx.Transaction.Outputs.Count;
+
+			var ownInputCount = tx.WalletInputs.Count;
+			var ownOutputCount = tx.WalletOutputs.Count;
+
+			if (ownInputCount == 0)
+			{
+				AnalyzeReceive(tx);
+			}
+			else if (inputCount == ownInputCount)
+			{
+				if (outputCount == ownOutputCount)
+				{
+					AnalyzeSelfSpend(tx);
+				}
+				else
+				{
+					AnalyzeNormalSpend(tx);
+				}
+			}
+			else
+			{
+				AnalyzeCoinjoin(tx);
+			}
+
+			AnalyzeClusters(tx);
+		}
+
+		private void AnalyzeCoinjoin(SmartTransaction tx)
+		{
+			AnalyzeWalletInputs(tx, out HashSet<HdPubKey> distinctWalletInputPubKeys, out int newInputAnonset);
+
+			var indistinguishableWalletOutputs = tx
+				.WalletOutputs.GroupBy(x => x.Amount)
+				.ToDictionary(x => x.Key, y => y.Count());
+
+			foreach (var newCoin in tx.WalletOutputs.ToArray())
 			{
-				// Calculate anonymity sets.
 				// Get the anonymity set of i-th output in the transaction.
 				var anonset = tx.Transaction.GetAnonymitySet(newCoin.Index);
 
-				// If we provided inputs to the transaction.
-				if (tx.WalletInputs.Any())
+				// Don't create many anonset if we've provided a lot of equal outputs.
+				anonset -= indistinguishableWalletOutputs[newCoin.Amount] - 1;
+
+				// We should only consider ourselves once in the anonset.
+				anonset += newInputAnonset - 1;
+
+				HdPubKey hdPubKey = newCoin.HdPubKey;
+				if (hdPubKey.AnonymitySet == HdPubKey.DefaultHighAnonymitySet)
```

Comment:
> what happens when a wallet is recovered from seed?

Works perfectly. Our blockchain analyser is stateless, there's no serialization of anonsets that can ruin things.

> what about if people export keys and issue some manual transactions with some other wallet software?

For the same reason, it works perfectly.

> what about broadcast but unconfirmed transactions? those would not be available to recovering wallets, so should they be taken into account? arguably yes because they were publicly observable, and can only decrease the scores.

They are taken into account. There's one edge case issue with that though that I'm not planning to solve:

1. Assume a tx is broadcast that address reused.
2. Our reused address's anonset is set properly.
3. Tx falls out of the mempool/RBF changes is or something funky happens.

In this case we're going to work properly until we're running, but after restart only the last state will be taken into consideration.

Unfortunately every clean way to solve this is many man month work or more.

> i think there's an implicit dependency on the blockchain order to produce stable estimates

Yes, there's an implicit dependency on blockchain order for all kinds of stuff. Long time ago we tried to speed up stuff by examining transactions and transaction outputs in parallel, but there's no way to have a correct wallet state like that. You can take it granted that analization will happen in order: block, tx index and mempool first seen.

>  if i'm not mistaken, by knowing more about total reuse a recovered wallet might re-estimate the anonymity set sizes of historical outputs more conservatively than a wallet that has not yet experienced any address reuse.

No, it gives the same estimate every time as far as I can comprehend it.

> for example, suppose a dust attack occurs on your coinjoined outputs, and you later choose to coinjoin the dust anyway, or donate it somewhere. when recovering the wallets is the privacy of the past coinjoins lost retrospectively?

One cannot see/touch coins under the dust threshold. Manual dust threshold change requires wallet restart. If the wallet is restarted these anonsets are recalculated (no serialization happens.) So those examples aren't triggering change on recovery, only on explicit dust threshold change.

> i would say the priority should be to the accuracy of estimates for unspent coins

ACK